pkg/connector/ble/ble.go

```diff
@@ -248,54 +320,83 @@ func tryToConnect(ctx context.Context, vin string, target Advertisement) (*Conne
 		}
 	}
 
-	if target.LocalName() != localName {
-		return nil, false, fmt.Errorf("ble: beacon with unexpected local name: '%s'", target.LocalName())
-	}
-
-	if !target.Connectable() {
-		return nil, false, ErrMaxConnectionsExceeded
+	if target.LocalName != localName {
+		return nil, false, fmt.Errorf("ble: beacon with unexpected local name: '%s'", target.LocalName)
 	}
 
-	log.Debug("Dialing to %s (%s)...", target.Addr(), localName)
-
-	client, err := ble.Dial(ctx, target.Addr())
-	if err != nil {
-		return nil, true, fmt.Errorf("ble: failed to dial for %s (%s): %s", vin, localName, err)
+	log.Debug("Connecting to %s (%s)...", target.Address.String(), localName)
+
+	// FIXME: This is a workaround for the fact that bluetooth library doesn't
+	// support context handling. While not a big issue, it will be good to
+	// have this fixed upstream. Also applies to the DiscoverServices and
+	// DiscoverCharacteristics calls below. See:
+	// https://github.com/tinygo-org/bluetooth/issues/339
+	deviceCh := make(chan bluetooth.Device)
+	errorCh := make(chan error)
+	connectionCancelled := false
+	go func() {
+		params := bluetooth.ConnectionParams{}
+		// If a deadline is set, use it as the connection timeout
+		// else this go routine will expire after the default timeout.
+		if deadline, ok := ctx.Deadline(); ok {
+			params.ConnectionTimeout = bluetooth.NewDuration(time.Until(deadline))
+		}
+		device, err := adapter.Connect(target.Address, params)
+		if err != nil && !connectionCancelled {
+			errorCh <- err
```

Comment:
I think there's still a race condition here; if we enter into this `if` branch, and then the main thread sets `connectionCancelled = true`, we'll be blocked trying to write to `errorCh` and no one will ever unblock us by reading from it. Similarly, there's a race with the `else if` branch. We could solve both of these problems by making the `errorCh` and `deviceCh` channels buffered, which will allow the goroutine to write to them even if no one is listening:


```
	deviceCh := make(chan bluetooth.Device, 1)
	errorCh := make(chan error, 1)
```